src/hotspot/share/opto/vector.cpp

```diff
@@ -424,7 +486,83 @@ Node* PhaseVector::expand_vbox_alloc_node(VectorBoxAllocateNode* vbox_alloc,
   return vec_obj;
 }
 
-void PhaseVector::expand_vunbox_node(VectorUnboxNode* vec_unbox) {
+Node* PhaseVector::expand_vbox_alloc_node(VectorBoxAllocateNode* vbox_alloc,
+                                          Node* value,
+                                          const TypeInstPtr* box_type,
+                                          const TypeVect* vect_type) {
+  ciInstanceKlass* box_klass = box_type->instance_klass();
+  if (is_vector(box_klass)) {
+    return expand_vbox_alloc_node_vector(vbox_alloc, value, box_type, vect_type);
+  } else {
+    return expand_vbox_alloc_node_mask_shuffle(vbox_alloc, value, box_type, vect_type);
+  }
+}
+
+Node* PhaseVector::get_loaded_payload(VectorUnboxNode* vec_unbox) {
+   Node* obj = vec_unbox->obj();
+   while(obj->is_InlineType()) {
+      obj = obj->as_InlineType()->field_value(0);
+   }
+   if (obj->bottom_type()->isa_vect()) {
+     return obj;
+   }
+   return NULL;
+}
+
+void PhaseVector::expand_vunbox_node_vector(VectorUnboxNode* vec_unbox) {
+  if (vec_unbox->outcnt() > 0) {
+    GraphKit kit;
+    PhaseGVN& gvn = kit.gvn();
+
+    Node* vec_val_load = get_loaded_payload(vec_unbox);
+
+    if (vec_val_load == NULL) {
+      Node* obj = vec_unbox->obj();
+      const TypeInstPtr* tinst = gvn.type(obj)->isa_instptr();
+      assert(VectorSupport::is_vector(tinst->instance_klass()), "");
+      ciInlineKlass* box_klass = static_cast<ciInlineKlass*>(tinst->instance_klass());
+
+      const TypeVect* vt = vec_unbox->bottom_type()->is_vect();
+      BasicType bt = vt->element_basic_type();
+      int num_elem = vt->length();
+
+      assert(VectorSupport::is_vector(box_klass), "");
+      ciField* payload_field = box_klass->declared_nonstatic_field_at(0);
+      int offset = payload_field->offset();
+      if (!payload_field->is_flattened()) {
+        ciInlineKlass* payload_klass = static_cast<ciInlineKlass*>(payload_field->type());
+        assert(VectorSupport::is_vector_payload_mf(payload_klass), "");
+        ciField* mutifield = payload_klass->declared_nonstatic_field_at(0);
+        offset += mutifield->offset();
+      }
+
+      Node* mem = vec_unbox->mem();
+      Node* ctrl = vec_unbox->in(0);
+      Node* vec_adr = gvn.transform(kit.basic_plus_adr(obj, offset));
+      const TypePtr *adr_type = gvn.type(vec_adr)->isa_ptr();
+
+      vec_val_load = LoadVectorNode::make(0,
+                                          ctrl,
+                                          mem,
+                                          vec_adr,
+                                          adr_type,
+                                          num_elem,
+                                          bt);
+      vec_val_load = gvn.transform(vec_val_load);
```

Comment:
Can simply use "vec_unbox->obj()->as_as_InlineType()->field_value(0)" ?

Reply:
This is a fallback code, which ideally should never get executed, i introduced it in first version of patch, but its not needed anymore.